File: src/main/java/com/github/magif1712/smarter_touhou_maids/features/smarter/agent/reflex_arc_system_agent/ai/process_ai/process/urana_process/fittable_mapper/nn/original_cnn/containers/io/CnnIO.java
package com.github.magif1712.smarter_touhou_maids.features.smarter.agent.reflex_arc_system_agent.ai.process_ai.process.urana_process.fittable_mapper.nn.original_cnn.containers.io;

import com.github.magif1712.smarter_touhou_maids.core.containers.vector.FloatVector;
import com.github.magif1712.smarter_touhou_maids.features.smarter.agent.reflex_arc_system_agent.ai.process_ai.process.urana_process.fittable_mapper.nn.original_cnn.containers.io.value.CnnInputVector;
import com.github.magif1712.smarter_touhou_maids.features.smarter.agent.reflex_arc_system_agent.ai.process_ai.process.urana_process.fittable_mapper.nn.original_cnn.containers.io.value.CnnOutputVector;

/**
 * CNN IO：把"输入向量 + 输出向量"这个不实在对偶，实在化为一个对象（真善美第4条）。
 * {@code getA0/getA1} 直接返回 underlying {@link FloatVector} 供 ops 桥接层提取句柄。
 * {@code close()} 释放 input + output。
 */
public class CnnIO implements AutoCloseable {
    private final CnnInputVector input;
    private final CnnOutputVector output;

    public CnnIO(int sizeA0, int sizeA1) {
        this.input = new CnnInputVector(sizeA0);
        this.output = new CnnOutputVector(sizeA1);
    }

    public CnnInputVector getInput() {
        return input;
    }

    public CnnOutputVector getOutput() {
        return output;
    }

    public FloatVector getA0() {
        return input.getVector();
    }

    public FloatVector getA1() {
        return output.getVector();
    }

    @Override
    public void close() {
        input.close();
        output.close();
    }
}

File: Document/Python-style pseudocode document/urana_process/fittable_mapper/nn/original_cnn/containers/io/cnn_io.py
from core import FloatVector
from urana_process.fittable_mapper.nn.original_cnn.containers.io.value.cnn_input_vector import CnnInputVector
from urana_process.fittable_mapper.nn.original_cnn.containers.io.value.cnn_output_vector import CnnOutputVector


class CnnIO:
    def __init__(self, sizeA0, sizeA1):
        self.input = CnnInputVector(sizeA0)
        self.output = CnnOutputVector(sizeA1)

    def getInput(self):
        return self.input

    def getOutput(self):
        return self.output

    def getA0(self):
        return self.input.getVector()

    def getA1(self):
        return self.output.getVector()

    def close(self): ...
